--- ./tmp_merged_foocanvas/configure.in	2010-03-12 18:06:32.277000000 +0000
+++ ./tmp_zmap_foocanvas/configure.in	2010-03-12 18:06:23.192000000 +0000
@@ -3,7 +3,8 @@
 AC_CONFIG_SRCDIR([libfoocanvas/foo-canvas.c])
 
 AC_CONFIG_HEADERS([config.h])
-AM_INIT_AUTOMAKE([dist-bzip2])
+dnl [dist-bz2] removed from the AM_INIT_AUTOMAKE()
+AM_INIT_AUTOMAKE([])
 
 AM_MAINTAINER_MODE
 
@@ -53,7 +54,8 @@
   AC_SUBST(HTML_DIR, $with_html_dir)
 fi
 
-GTK_DOC_CHECK([1.0])
+dnl MAC OS X doesn't like this... (requires gtk-doc, which requires a million other packages)
+dnl GTK_DOC_CHECK([1.0])
 
 AC_CONFIG_FILES([
 Makefile
