<h2>BAM file data (short reads)</h2>

<fieldset><legend>Use by ZMap</legend>
<p>It is thought that the volume of data is too great for direct use by ZMap as the time needed to load and display woud be impractical.  Instead ZMap will load summary data and display data volumes in fixed size bins over the sequence in view.  The actual data will be displayed by Blixem and requested over shorter regions defined by the mark.
</p>
<p>In the short term we do not expect to be able to handle summary data and instead provide a menu driven interface to run Blixem.  This will involve adding a 'Blixem Short Reads' option to the column menu for each configured short-reads data sources.  Later on when summary data is available the summary columns can be Blixem'ed directly.
</p>
<p>By default Blixem operates with a region of 200k bases centred on the mouse (which can be set by the Edit/Preferences main menu item) but for short reads data this will be different: the user must set the mark and if not set the menu options for short reads data will not appear, or if selected will pop up a reminder message.
</p>
</fieldset>

<fieldset><legend>Configuration</legend>
<p>
The major problem to solve is one of passing round information regarding what data to request.  Otterlace has to supply enough configuration information to ZMap and Zmap has to pass on this plus sequence information to Blixem.
</p>
<p>BAM is a file format not a data source and we need to be able to specify multiple data sources in the BAM format and present these to the user with relevant names.
</p>

<h3>Sequence definition</h3>
<p>There are currently a number of ways that ZMap is asked to load and display data and these operate by some implicit protocols.  It may be appropriate to redefine some of this to avoid duplication in Blixem configuration.
</p>

<p>ZMap can be started up without specifying a 'default sequence' and this results in the main window (not normally visible to Otterlace users) being ready to accept sequence requests, and no view window being displayed.  Views can be created by requesting this from the main window (the ZMap Manager) or via an XRemote request from Otterlace, (both of which reuqire a sequence name and start and end coordinates) but note that currently there is no clear linkage between a view and the data sources or request protocol used to request the data.  This has implications for <a href="Design/notes/standalone.shtml">standalone operation</a> and currently the ability to specify different sequences is limited to operation with ACEDB.  Note that ZMap does not provide this kind of function with pipe servers.
</p>

<p> ZMap will use the sequence name of the current view and the requested start and end coordinates (specified by the mark) as the source of this data and pass these directly to Blixem.  Blixem will call a script provided by Anacode to generate a GFF file containing BAM file sourced data and pass this data to the script via command line arguments.
</p>


<h3>[Blixem] configuration - new options</h3>
<h4>Configuration file</h4>
<p>
The blixem configuration file <b>blixemrc</b> will be ammended to include a stanza for short reads data, which will include a script with optional arguments, not including the sequence name or start or end coordinates. Refer <a href="Design/notes/standalone.shtml">here</a> for some relevant comments regarding other pipe-server arguments which may affect this configuration in future.
</p>
<p>
The config file already specifies the default pfetch mode for our current data type (EMBL sequences).  This fetch mode relates to fetching missing data for sequences that Blixem already knows about (i.e. those in its &lt;datafile&gt; input file).  The new fetch method will fetch all sequences within a particular region from a particular source, without any prior knowledge of what those sequences are.  We could in the future want additional fetch methods for different data types that we don't yet know about, which may fetch sequences or just sequence data.
</p>
<p>The following suggestion for a config file format is an attempt to be as flexible as possible in these requirements:
<pre>
[blixem]
default-fetch = pfetch-socket  # default fetch method to use when there is no data type given for an item in the GFF file

[embl]                         # config options for the embl data type
fetch = pfetch-socket

[short-read]                   # config options for the short-read data type
fetch = bam-fetch

[pfetch-socket]                # config options for this particular fetch method
pfetch-mode = socket
port = 22400
node = pfetch.sanger.ac.uk

[bam-fetch]                    # config options for this particular fetch method
script = /software/anacode/bin/scripts/get_seqdata.pl
arguments = stuff
</pre>
</p>
<p>The GFF file would then look something like:
<pre>
##gff-version 3
##sequence-region chr4-04 210623 364887   # the reference sequence region that is passed to blixem
chr4-04	source1	region	215000	300000	0.000000	.	.	dataType=short-read      # the coords here specify the region to fetch short-reads for
chr4-04	source2	region	215000	300000	0.000000	.	.	dataType=short-read      # optionally fetch short-reads for a second source
chr4-04_210623-364887	vertebrate_mRNA	nucleotide_match	212438	212568	71.000000	+	.	Target=AK075066.1 125 255 +;percentID=77.1;dataType=embl
chr4-04	vertebrate_mRNA	nucleotide_match	211806	212636	544.000000	+	.	Target=AK123078.1 831 1683 +;percentID=84;  # no dataType tag => use default-fetch
...
</pre>
Here, the source column (say, "source1") is the data that would be passed to the bam-fetch script's --analysis option.  If this doesn't make sense as the source column, we could place it in another custom data tag instead (e.g., "dataSource=source1").
</p>
<h4>Command line arguments</h4>
<p>Note that Blixem will need to pass chromosome coords to the bam-fetch script.  ZMap should pass everything to Blixem in chromosome coords plus the required offset (via the -O argument) that Blixem will need to use to display its coords in the same coordinate system as ZMap.</p>
<p>
All other options relating to short-reads data will be passed in the GFF file and no further command line arguments are neeeded.
</p>

<h3>[ZMap] stanza - new options</h3>
<p> As ZMap has to use this to create a menu it will be better to define this globally rather than via Blixem.

<table border="1"  cellpadding="3" cellspacing="3">
<tr> <td> SeqData</td> <td>A list of columns or featuresets containing SR data<br />
As for DNA and protein featuresets we try to match both columns<br /> and featuresets to allow simple configuration.<br /> (Note: each column/featureset has options for short names and more lengthy descriptions). <br />
Refer <a href="Design/notes/standalone.shtml">here</a> for some relevant comments regarding extra ZMap config options
 </td> </tr>
<tr> <td> </td> <td> </td> </tr>
</table>
</p>

</fieldset>


<fieldset><legend>Interface to the Request script for Short Reads data</legend>
<p>Blixem will find the script path and fixed arguments from a stanza in its configuration file 'blixemrc'.
Other arguments will be appended as follows:
<table border="1"  cellpadding="3" cellspacing="3">
<tr> <td> --analysis</td> <td>  source name of the data we wish to requesty</td> </tr>
<tr> <td> --gff_sequence</td> <td> the name of the sequence eg 'chr6-18'</td> </tr>
<tr> <td> --start</td> <td> request start coordinate</td> </tr>
<tr> <td> --end</td> <td> request end coordinate</td> </tr>
<tr> <td> --dataset</td> <td> where to get the sequence from eg 'human' <br /><b>Note</b> in the short term this should appear in the fixed arguments list in blixemrc.</td> </tr>
<tr> <td> --gff_file</td> <td> where to put the output</td> </tr>
</table>

</p>

</fieldset>
