<h2>BAM file data (short reads)</h2>

<fieldset><legend>Use by ZMap</legend>
<p>It is thought that the volume of data is too great for direct use by ZMap as the time needed to load and display woud be impractical.  Instead ZMap will load summary data and display data volumes in fixed size bins over the sequence in view.  The actual data will be displayed by Blixem and requested over shorter regions defined by the mark.
</p>
<p>In the short term we do not expect to be able to handle summary data and instead provide a menu driven interface to run Blixem.  This will involve adding a 'Blixem Short Reads' option to the column menu for each configured short-reads data sources.  Later on when summary data is available the summary columns can be Blixem'ed directly.  This menu option will be interpreted as if the user clicked on a summary column and not a feature.
</p>
<p>By default Blixem operates with a region of 200k bases centred on the mouse (which can be set by the Edit/Preferences main menu item) but for short reads data this will be different: the user must set the mark and if not set the menu options for short reads data will not appear.
</p>
</fieldset>

<fieldset><legend>Configuration</legend>
<p>
The major problem to solve is one of passing round information regarding what data to request.  Otterlace has to supply enough configuration information to ZMap and Zmap has to pass on this plus sequence information to Blixem.
</p>
<p>BAM is a file format not a data source and we need to be able to specify multiple data sources in the BAM format and present these to the user with relevant names.
</p>

<h3>Sequence definition</h3>
<p>There are currently a number of ways that ZMap is asked to load and display data and these operate by some implicit protocols.  It may be appropriate to redefine some of this to avoid duplication in Blixem configuration.
</p>

<p>ZMap can be started up without specifying a 'default sequence' and this results in the main window (not normally visible to Otterlace users) being ready to accept sequence requests, and no view window being displayed.  Views can be created by requesting this from the main window (the ZMap Manager) or via an XRemote request from Otterlace, (both of which reuqire a sequence name and start and end coordinates) but note that currently there is no clear linkage between a view and the data sources or request protocol used to request the data.  This has implications for <a href="Design/notes/standalone.shtml">standalone operation</a> and currently the ability to specify different sequences is limited to operation with ACEDB.  Note that ZMap does not provide this kind of function with pipe servers.
</p>

<p> ZMap will use the sequence name of the current view and the requested start and end coordinates (specified by the mark) as the source of this data and pass these directly to Blixem.  Blixem will call a script provided by Anacode to generate a GFF file containing BAM file sourced data and pass this data to the script via command line arguments.
</p>


<h3>[Blixem] configuration - new options</h3>
<h4>Configuration file</h4>
<p>
The blixem configuration file <b>blixemrc</b> will be ammended to include a stanza for fetching short-read data, which will include a script with optional arguments, not including the sequence name or start or end coordinates. Refer <a href="Design/notes/standalone.shtml">here</a> for some relevant comments regarding other pipe-server arguments which may affect this configuration in future.
</p>
<p>
The config file already specifies the default pfetch mode for our current data type (EMBL sequences).  This fetch mode relates to fetching missing data for sequences that Blixem already knows about (i.e. those in its &lt;datafile&gt; input file).  The new fetch method will fetch all sequences within a particular region from a particular source, without any prior knowledge of what those sequences are.  We could in the future want additional fetch methods for different data types that we don't yet know about, which may fetch sequences or just sequence data.
</p>
<p>The following suggestion for a config file format is an attempt to be as flexible as possible in these requirements. See the <a href="http://mediawiki.internal.sanger.ac.uk/wiki/index.php/Blixem_Config_file">wiki page</a> for more info about the Blixem config file.
<pre>
[blixem]
bulk-fetch = pfetch-socket     # default fetch method to use when there is no data type given for an item in the GFF file
user-fetch = pfetch-http       # optionally, this can be used to specify a different fetch method for when the user double-clicks a sequence

[embl]                         # config options for the embl data type
bulk-fetch = pfetch-socket

[short-read]                   # config options for the short-read data type
bulk-fetch = region-fetch

[region-fetch]                 # config options for this particular fetch method
script = /nfs/users/nfs_j/jh13/work/bam/bam_get


...
</pre>
</p>
<p>Assuming that we have two BAM sources configured (in ZMap) as:
<pre>
[ZMap]
seq-data=14dpf;Liver
</pre>
the Blixem configuration file will include new stanzas for each source of BAM data in a format as follows:
<pre>
[14dpf]
args=-bam_path=path_to_data_file =bam_cs=NBCIM37
name=14dpf
description=14 days post fertilisation

[Liver]
args=-bam_path=path_to_data_file =bam_cs=NBCIM37
description=Adult Liver
etc
</pre>
and Blixem with know to read the relevant stanza by referring to the GFF file provided by ZMap with a single line for each short-reads request.  The name of each stanza will be exactly as specifed in the source field.  For Blixem the name and description items are optional.
</p>
</p>
<p>Note that this implies that the blixem configuration file must not be generated for each invocation of Blixem if the dataset changes.
</p>

<p>The GFF file would then look something like:
<pre>
##gff-version 3
##sequence-region chr4-04 210623 364887   # the reference sequence region that is passed to blixem
chr4-04	source1	region	215000	300000	0.000000	.	.	dataType=short-read      # the coords here specify the region to fetch short-reads for
chr4-04	source2	region	215000	300000	0.000000	.	.	dataType=short-read      # optionally fetch short-reads for a second source
chr4-04	vertebrate_mRNA	nucleotide_match	212438	212568	71.000000	+	.	Target=AK075066.1 125 255 +;percentID=77.1;dataType=embl
chr4-04	vertebrate_mRNA	nucleotide_match	211806	212636	544.000000	+	.	Target=AK123078.1 831 1683 +;percentID=84;  # no dataType tag => use default-fetch
</pre>

</p>
<h4>Command line arguments</h4>
<p>Note that Blixem will need to pass chromosome coords to the bam-fetch script.  ZMap should pass everything to Blixem in chromosome coords plus the required offset (via the -O argument) that Blixem will need to use to display its coords in the same coordinate system as ZMap.</p>
<p>
All other options relating to short-reads data will be passed in the GFF file and Blixem configuration and no further command line arguments are needed.
</p>

<h3>[ZMap] stanza - new options</h3>
<p>

<table border="1"  cellpadding="3" cellspacing="3">
<tr> <td> seq-data</td> <td>A list of featuresets containing SR data<br />
As for DNA and protein featuresets we try to match both columns<br /> and featuresets to allow simple configuration.<br /> (Note: each column/featureset has options for short names and more lengthy descriptions). <br />
Refer <a href="Design/notes/standalone.shtml">here</a> for some relevant comments regarding extra ZMap config options
 </td> </tr>
</table>

ZMap will hold this list in the View's context_map (accessable to the Window code) and it will be used in the short term to construct RC menus for Blixem.
</p>

<p>The [ZMap] stanza will also include:

<table border="1"  cellpadding="3" cellspacing="3">
<tr> <td> dataset</td> <td> eg 'human' </td> </tr>
<tr> <td> sequence</td> <td> eg 'chr14-10'</td> </tr>
</table>
</p>

</fieldset>



<fieldset><legend>Interface to the Request script for Short Reads data</legend>
<p>Blixem will find the script path and fixed arguments from a stanza in its configuration file 'blixemrc'.
Other arguments will be appended as follows:
<table border="1"  cellpadding="3" cellspacing="3">
<tr> <td> -chr</td> <td> the name of the reference sequence eg 'chr6-18'</td> </tr>
<tr> <td> -gff_seqname</td> <td> (alternative format of '-chr' for use by ZMap)</td> </tr>
<tr> <td> -start</td> <td> request start coordinate</td> </tr>
<tr> <td> -end</td> <td> request end coordinate</td> </tr>
<tr> <td> -bam_path</td> <td> path to the BAM file </td></tr>
<tr> <td> -bam_cs</td> <td> coordinate system of the BAM file, probably NCBIM37</td></tr>
</table>

</p>
<p>In the near future ZMap will also be able to request data from this script and all script configuration will be done via the existing pipe server mechanism.
</p>
<h3>ZMap configuration for BAM data requests</h3>
<p>As specified above
<pre>
[ZMap]
seq-data=14dpf;Liver
</pre>
will be used to specify BAM data that may be used requested by Blixem.
ZMap will in future be able to display coverage data for BAM datasets and also request the real data from the mark by R-clicking on a coverage column..
Each set of BAM data and coverage data will be named as distinct featuresets and these may be requested via the existing pipe server mechanism - either by a request on startup or via the columns dialog or by request from otterlace.
We expect coverage data to be displayed by default and the users to request sub-sequences of BAM data by using the RC menu on a coverage column (although there is a request for a special pane on the load columns dialog).
</p>
<p>
Each BAM pipe server will support the relevant ZMap featureset, and coverage data will link to real data via another stanza in the ZMap configuration:
<pre>
[Coverage]
14dpf_coverage=14dpf		# (coverage = real data)
</pre>
and right clicking on a coverage column will display an option to display the real data column. (This will replace the inclusion of all BAM data sets in the RC menu as at present). Whether or not this causes the columns to switch or display together is not relevant here.  Existing ZMap configuration (eg [featureset-description]) will be used to specify display names and helpful information as at present.
</p>
<p>Referring to the <a href="Design/notes/RNAseq.shtml">request</a> for a new pane on the load columns dialog, the existance of a [Coverage] mapping may be used to include featuresets in this page or not, and also the existing load columns pane.
</p>

</fieldset>


<a name="blixem_bam"></a>
<fieldset><legend>Passing short reads dataset choice to blixem</legend>
<h3>Window Menus</h3>
<p>Depending on where the user clicks a differnt choice of Blixem menu option appear and in the short term all cases will include access to short reads data as configured, but only if the mark is set. It would be preferable to display these as disabled/ greyed out  but this option has not been programmed and would require extra work to include.
</p>
<p>In <b>zmapWindowMenus.c</b> an anonymous enum of BLIX_ options has been defined and this is used to specify what kind of homology to fetch (DNA, peptide, column or feature(s) and we can add a new open BLIX_SEQ.  Eventually we aim to interface by the user clicking on the desired column but in the short term BLIX_SEQ will be used as the base of a range of values referring to a list of featuresets defined in the ZMap config, which is stored in the view's context map and is accessable to the Window code. (<b>view|window-&gt;context_map.seq_data_featuresets</b>).  When short reads data can be displayed ion ZMap then this can revert to a single value.
</p>
<p>This data is copied to a data structure <b>ZMapWindowCallbackCommandAlign</b> by <b>zmapWindowFeatureFuncs.c/zmapWindowCallBlixemOnPos()</b>.  There is another enum <b>ZMapWindowAlignSetType</b> in <b>zmapWindow.h</b> which defines the same options as the BLIX_ stuff but is accessable by other modules.  We map BLIX_ to ZMAPWINDOW_ALIGNCMD_ and at the same time insert the appropriate featureset as chosen from the menu as if the user clicked on a column.
</p>

<h3>Call Blixem</h3>
<p>Blixem expect to receive a list of features in a GFF file and for short reads data the features are not known and we give a 'source'  feature instead corresponding to the featureset, which we will pass through as text (the name).
</p>
<p>Th interface to the <b>zmapViewCallBlixem()</b> code is via the function<b>doBlixemCmd()</b> which is given a
<b>ZMapWindowCallbackCommandAlign</b> structure and calls zmapViewCallBlixem() passing as arguments each of the members of that structure. doBlixemCommand() copies these arguments to another data structure <b>blixemDataStruct</b> which is private to zmapViewCallBlixem.c.
</p>

</fieldset>
