</div>
<div style="margin-right:5em;margin-left:11em;">
<h2>Coordinate systems and configuration</h2>
<fieldset><legend>Coordinate systems for generic daatbases</legend>
<h3>What's the problem?</h3>
<p>
It is common (especially in model organism databases) to use chromosome coordinates to refer to sequence information but many databases exist which are based on fragments eg BAC (clones), and these are typically based from 1.  ZMap implements a mapping process to define what coordinates are presented to the user and what coordinates are used when referring to the underlying database.
</p>
<p> ZMap will display coordinates on the Scale bar and on the Ruler as block relative - the first base in any block will be 1 and the last will be the length of the block + 1.  The Ruler will also display coordinates in the parent span if these differ from block relative.  The Status bar will show coordinates as block relative (which will be consistent with the scale bar). Currently there is no way to display parent (eg chromosome) coordinates in the status bar.
</p>
<p>ZMap stores sequence and feature data in a <b>FeatureContext</b> which contains typically a single <b>Align</b> which contains a single contiguous <b>Block</b> of features.  There is provision to have multiple aligns and blocks and in this case the blocks of features could be mapped from any genomic region, for example to model genome rearrangements.   Note however that each align contains blocks from the same strand. Each align may be flagged as being in the reverse strand relative to the parent sequence.
</p>
<p>Requests for data from servers can be made for single blocks only.
</p>
<p>Much of the top level interface to ZMap specifies a sequence as start to end, which implies a single align and block, and at present multiple blocks and aligns are not supported.
</p>


<h3>Data used to perform coordinate mapping</h3>
<h4>ZMapFeatureContext->parent_span</h4>
<p>
This defines the genomic region the sequence(s) being viewed are extracted from. Note that at present it is assumed that this is a single contiguous sequence of DNA (eg a single chromosome or a database holding a sequence based on some other coordinate system.  If more than one chromosome is to be represented then it is necessary to concatenate these into a single sequence; this will be reviewed in future and is likely to change.  Parent span coordinates are always forward strand.
<pre>
typedef struct
{
  Coord x1, x2 ;
} ZMapSpanStruct, *ZMapSpan ;

(in ZMapFeatureContextStruct)
ZMapSpanStruct parent_span;
</pre>
</p>

<h4>ZMapFeatureAlignment->seq_to_parent</h4>
<p>This defines the genomic region within the parent span that a series of blocks cover.  It contains the subsection of the parent span that is included, and the coordinates that are used to refer to the data being viewed internally by ZMap. There is a flag to say whether or not the region is reversed (ie reverse strand relative to the parent span)
<pre>
typedef struct
{
      /* NOTE even if reversed coords are as start < end */
  ZMapSpanStruct parent;          /* start/end in parent span (context) */
  ZMapSpanStruct align;             /* start,end in align, aka child seq */
  gboolean reversed;
} ZMapMapAlignStruct, *ZMapMapAlign ;


(in ZMapFeatureAlignmentStruct)
  ZMapMapBlockStruct sequence_to_parent ;
</pre>
</p>

<h4>ZMapFeatureBlock->block_to_sequence</h4>
<p>This defines the part of the containing align that is covered by a block of features. It is simply start and end coordinates and it is assumed that it is on the same strand as the containing align.
<pre>
(in ZMapFeatureBlockStruct)
ZMapSpanStruct block_to_sequence ;
</pre>
</p>

<h4>Overall structure</h4>
<p>Some examples, note that we do not limit ZMap to these two options)
</p>

<p>If the parent sequence and requested data are based from 1  (ACEDB model):
<pre>
  1  S                   E    X      chromosome bases
  =============================      ZMapFeatureContext.parent_span.x1,x2 = 1,X

     1                   Y           our sequence
     =====================           ZMapFeatureAlign.sequence_to_parent
                                       sequence.x1,x2 = S,E
                                       block.x1,x2    = 1,Y

        BS        BE                 displayed block
        ============                 ZMapFeatureBlockStruct.block_to_sequence.x1,x2 = BS,BE
</pre>
<p>If the parent sequence and requested data are chromosome coordinates (Otterlace model):
<pre>
  1  S                   E    X      chromosome bases
  =============================      ZMapFeatureContext.parent_span.x1,x2 = 1,X

     S                   E           our sequence
     =====================           ZMapFeatureAlign.sequence_to_parent
                                       sequence.x1,x2 = S,E
                                       block.x1,x2    = S,E

        BS        BE                 displayed block
        ============                 ZMapFeatureBlockStruct.block_to_sequence.x1,x2 = BS,BE
</pre>
</p>

<h3>Mapping versus Alignment</h3>
<p>
When mapping a genomic sequence to an underlying database both forward and reverse strands are mapped together.  This is different from alignment which is where a sequence on one strand of a DNA segment is aligned with a sequence on another region and strand (although the strands may in fact be the same).
</p>
<p>Note that there are two different data structures defined in <b>zmapFeature.h</b>: <b>ZMapAlignBlock</b> which is used to process gapped alignments, and <b>ZMapMapAlign</b> which is used to specify the mapping between Align and parent sequence.
</p>
</fieldset>


<fieldset><legend>Chromosome and ZMap coordinates</legend>
<p>Chromosome coordinates are useful when dealing with external sources and people but are unwieldy when annotating and ZMap has traditionally held features based on slice coordinates ie based from 1.
</p>
<p>The user will always be presented with ZMap coordinates in status widgets and the ruler also provides a tooltip with the zmap coordinate and if available the corresponding chromosome coordinate.
</p>
<p>Traditionally the chromosome coordinates have been derived from the sequence name (eg 'chr3-18_123124234-234242342').
</p>

<h3>New configuration options</h3>
<p>As a first step towards specifying the regiojn of interest in ZMap configuration we will provide the following options:
<pre>
[ZMap]
start=12324124
end=234242234
csname=chromosome
csver=Otter
</pre>
and initially any values other than chromosome and Otter will be invalid.
</p>
<p>Subsequently various other parameters currently used in pipe server URL's will be controlled by extra option in [Zmap].
</p>

<h3>Use with otterlace</h3>
<p>dev_otterlace (and evetually test_otterlace and otterlace) will accept an environment variable to choose between chromosome coordinates and 1-based coordinates.  This will control whether of not start and end is configured in ZMap and also whether pipe server scripts rebase thier coordinates from 1. ACEDB must be configured seperately.
</p>


<h3>Handling chromosome coordinates without new configuration</h3>
<p>If available, the chromosome coordinates in the sequence name in the ZMap config will be used to present chromosome coordinates on request by the user (eg in the ruler tooltip).
Pipe servers and ACEDB should provide GFF data based from 1 and requests for data will be based from 1.
</p>

<h3>Handling chromosome coordinates with new configuration</h3>
<p>
ZMap will store features with chromosome coordinates and Pipe servers and ACEDB must provide these in their GFF output.  Coordinates presented to the user will be adjusted to be based from 1.  Requests for data (eg load from mark) will be in chromosome coordinates. <b>NB</b> ACEDB must be configured with extra data to facilitate this.
</p>

<h3>Implications for ZMap</h3>
<p>Data requests will always be in the native coordinate system.
</p>
<p>The scale display must be adjusted when choromosome coords are configured.
</p>

<fieldset>
