diff -ru gtk+-2.10.14/acconfig.h gtk+-2.10.15/acconfig.h
--- acconfig.h	2007-07-16 20:46:25.000000000 +0100
+++ acconfig.h	2007-11-15 12:57:09.000000000 +0000
@@ -70,6 +70,8 @@
 
 #undef HAVE_CUPS_API_1_2
 
+#undef HAVE_CUPS_API_1_3
+
 /* #undef PACKAGE */
 /* #undef VERSION */
 
diff -ru gtk+-2.10.14/configure.in gtk+-2.10.15/configure.in
--- configure.in	2007-07-16 20:46:25.000000000 +0100
+++ configure.in	2007-11-15 12:56:59.000000000 +0000
@@ -1565,6 +1565,11 @@
     AC_DEFINE(HAVE_CUPS_API_1_2)
   fi
 
+  if test $CUPS_API_MAJOR -gt 1 -o \
+          $CUPS_API_MAJOR -eq 1 -a $CUPS_API_MINOR -ge 3; then
+    AC_DEFINE(HAVE_CUPS_API_1_3)
+  fi
+
   AC_SUBST(CUPS_API_MAJOR)
   AC_SUBST(CUPS_API_MINOR)
   AC_SUBST(CUPS_CFLAGS)
diff -ru gtk+-2.10.14/modules/printbackends/cups/gtkcupsutils.c gtk+-2.10.15/modules/printbackends/cups/gtkcupsutils.c
--- modules/printbackends/cups/gtkcupsutils.c	2007-11-15 12:51:24.000000000 +0000
+++ modules/printbackends/cups/gtkcupsutils.c	2007-11-15 13:03:32.000000000 +0000
@@ -1,3 +1,4 @@
+/*  Last edited: Nov 15 13:03 2007 (roy) */
 /* GTK - The GIMP Toolkit
  * gtkcupsutils.h: Statemachine implementation of POST and GET 
  * cup calls which can be used to create a non-blocking cups API
@@ -627,7 +628,9 @@
   httpClearFields(request->http);
   httpSetField(request->http, HTTP_FIELD_CONTENT_LENGTH, length);
   httpSetField(request->http, HTTP_FIELD_CONTENT_TYPE, "application/ipp");
-#ifdef HAVE_HTTP_AUTHSTRING
+#ifdef HAVE_CUPS_API_1_3
+  httpSetField(request->http, HTTP_FIELD_AUTHORIZATION, httpGetAuthString(request->http));
+#else
   httpSetField(request->http, HTTP_FIELD_AUTHORIZATION, request->http->authstring);
 #endif
 
@@ -966,7 +969,9 @@
     }
 
   httpClearFields(request->http);
-#ifdef HAVE_HTTP_AUTHSTRING
+#ifdef HAVE_CUPS_API_1_3
+  httpSetField(request->http, HTTP_FIELD_AUTHORIZATION, httpGetAuthString(request->http));
+#else
   httpSetField(request->http, HTTP_FIELD_AUTHORIZATION, request->http->authstring);
 #endif
 
