--- ./tmp_merged_foocanvas/libfoocanvas/foo-canvas-text.c	2010-03-16 13:12:15.885000000 +0000
+++ ./tmp_zmap_foocanvas/libfoocanvas/foo-canvas-text.c	2010-03-16 13:13:24.157000000 +0000
@@ -1,7 +1,7 @@
-/*  Last edited: Mar  8 10:25 2010 (edgrif) */
+/*  Last edited: Mar 16 13:13 2010 (edgrif) */
 /* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*- */
 /*
- * $Id: foo-canvas-text.c,v 1.1 2010-03-10 15:27:23 edgrif Exp $
+ * $Id: foo-canvas-text.c 59 2008-06-21 15:57:11Z jody $
  * Copyright (C) 1997, 1998, 1999, 2000 Free Software Foundation
  * All rights reserved.
  *
