--- ./tmp_merged_foocanvas/libfoocanvas/foo-canvas-polygon.c	2010-03-16 15:14:38.358000000 +0000
+++ ./tmp_zmap_foocanvas/libfoocanvas/foo-canvas-polygon.c	2010-03-16 15:14:36.903000000 +0000
@@ -1,3 +1,4 @@
+/*  Last edited: Mar  4 15:18 2010 (edgrif) */
 /*
  * Copyright (C) 1997, 1998, 1999, 2000 Free Software Foundation
  * All rights reserved.
@@ -253,7 +254,7 @@
 	/* Add outline width */
 
 	if (poly->width_pixels)
-		width = poly->width / poly->item.canvas->pixels_per_unit;
+		width = poly->width / poly->item.canvas->pixels_per_unit_x;
 	else
 		width = poly->width;
 
@@ -377,7 +378,7 @@
 	if (poly->width_pixels)
 		width = (int) poly->width;
 	else
-		width = (int) (poly->width * poly->item.canvas->pixels_per_unit + 0.5);
+		width = (int) (poly->width * poly->item.canvas->pixels_per_unit_x + 0.5);
 
 	gdk_gc_set_line_attributes (poly->outline_gc, width,
 				    GDK_LINE_SOLID, GDK_CAP_ROUND, GDK_JOIN_ROUND);
@@ -778,7 +779,7 @@
 
 	if (poly->outline_set) {
 		if (poly->width_pixels)
-			width = poly->width / item->canvas->pixels_per_unit;
+			width = poly->width / item->canvas->pixels_per_unit_x;
 		else
 			width = poly->width;
 
