--- ./tmp_merged_foocanvas/libfoocanvas/libfoocanvas.pc.in	2010-03-24 14:47:04.379000000 +0000
+++ ./tmp_zmap_foocanvas/libfoocanvas/libfoocanvas.pc.in	2010-03-24 14:47:01.523000000 +0000
@@ -3,9 +3,9 @@
 libdir=@libdir@
 includedir=@includedir@
 
-Name: libfoocanvas
-Description: libfoocanvas
+Name: libzmapfoocanvas
+Description: libzmapfoocanvas
 Requires: pango gtk+-2.0
 Version: @VERSION@
-Libs: -L${libdir} -lfoocanvas
-Cflags: -I${includedir}/libfoocanvas-1.0
+Libs: -L${libdir} -lzmapfoocanvas
+Cflags: -I${includedir}/libzmapfoocanvas-1.0
