--- ./tmp_merged_foocanvas/libfoocanvas/libfoocanvas.pc.in	2010-03-16 15:14:38.378000000 +0000
+++ ./tmp_zmap_foocanvas/libfoocanvas/libfoocanvas.pc.in	2010-03-16 15:14:36.982000000 +0000
@@ -3,9 +3,9 @@
 libdir=@libdir@
 includedir=@includedir@
 
-Name: libfoocanvas
-Description: libfoocanvas
+Name: libzmapfoocanvas
+Description: libzmapfoocanvas
 Requires: pango gtk+-2.0
 Version: @VERSION@
-Libs: -L${libdir} -lfoocanvas
+Libs: -L${libdir} -lzmapfoocanvas
 Cflags: -I${includedir}/libfoocanvas-1.0
